File: GoNorth/Game.cpp
//------------------------------------------------------------------------------------------------
// Game.cpp
//
// Implementation of CGame class
//------------------------------------------------------------------------------------------------

#include "stdafx.h"

namespace GoNorth
{
    CGame *CGame::s_pThis = nullptr;

    //------------------------------------------------------------------------------------------------
    DIRECTION GetReverseDirection(DIRECTION direction)
    {
        switch (direction)
        {
        case NORTH:
            return SOUTH;
        case NORTHEAST:
            return NORTHWEST;
        case EAST:
            return WEST;
        case SOUTHEAST:
            return NORTHWEST;
        case SOUTH:
            return NORTH;
        case SOUTHWEST:
            return NORTHEAST;
        case WEST:
            return EAST;
        case NORTHWEST:
            return SOUTHEAST;
        case UP:
            return DOWN;
        case DOWN:
            return UP;
        }
        return NUMDIRECTIONS; // ERROR
    }

    //------------------------------------------------------------------------------------------------
    CGame::CGame()
    {
        if (s_pThis)
        {
            // Singleton class only
            throw(std::bad_alloc());
        }

        m_logStream.Open("gonorth.log");

        // Register words
        m_dictionary.AddWord("Help",        WORD_HELP);
        m_dictionary.AddWord("Exit",        WORD_EXIT);
        m_dictionary.AddWord("Use",         WORD_USE);
        m_dictionary.AddWord("Go",          WORD_GO);
        m_dictionary.AddWord("Get",         WORD_GET);
        m_dictionary.AddWord("Drop",        WORD_DROP);
        m_dictionary.AddWord("Move",        WORD_MOVE);
        m_dictionary.AddWord("Look",        WORD_LOOK);
        m_dictionary.AddWord("North",       WORD_NORTH);
        m_dictionary.AddWord("Northeast",   WORD_NORTHEAST);
        m_dictionary.AddWord("East",        WORD_EAST);
        m_dictionary.AddWord("Southeast",   WORD_SOUTHEAST);
        m_dictionary.AddWord("South",       WORD_SOUTH);
        m_dictionary.AddWord("Southwest",   WORD_SOUTHWEST);
        m_dictionary.AddWord("West",        WORD_WEST);
        m_dictionary.AddWord("Northwest",   WORD_NORTHWEST);
        m_dictionary.AddWord("Up",          WORD_UP);
        m_dictionary.AddWord("Down",        WORD_DOWN);
        m_dictionary.AddWord("Yes",         WORD_YES);
        m_dictionary.AddWord("No",          WORD_NO);
        m_dictionary.AddWord("Inventory",   WORD_INVENTORY);

        // Aliases
        m_dictionary.AddWord("?",           WORD_HELP);
        m_dictionary.AddWord("Quit",        WORD_EXIT);
        m_dictionary.AddWord("Q",           WORD_EXIT);
        m_dictionary.AddWord("N",           WORD_NORTH);
        m_dictionary.AddWord("NE",          WORD_NORTHEAST);
        m_dictionary.AddWord("E",           WORD_EAST);
        m_dictionary.AddWord("SE",          WORD_SOUTHEAST);
        m_dictionary.AddWord("S",           WORD_SOUTH);
        m_dictionary.AddWord("SW",          WORD_SOUTHWEST);
        m_dictionary.AddWord("W",           WORD_WEST);
        m_dictionary.AddWord("NW",          WORD_DOWN);
        m_dictionary.AddWord("I",           WORD_INVENTORY);
        m_dictionary.AddWord("Y",           WORD_YES);
        m_dictionary.AddWord("N",           WORD_NO);

        // Add default player and select
        m_selectedPlayerId = MakePlayer<CPlayer>();

        // Set the default prompt string
        m_promptString = "> ";

        s_pThis = this;
    }

    //------------------------------------------------------------------------------------------------
    CGame::~CGame()
    {

    }

    //------------------------------------------------------------------------------------------------
    CLocation *CGame::GetLocation(LocationId locationId)
    {
        return locationId >= 0 && locationId < m_locations.size() ? m_locations[locationId].get() : nullptr;
    }

    //------------------------------------------------------------------------------------------------
    CPlayer *CGame::GetPlayer(PlayerId playerId)
    {
        return playerId >= 0 && playerId < m_players.size() ? m_players[playerId].get() : nullptr;
    }

    //------------------------------------------------------------------------------------------------
    void CGame::SetLocationNeighborId(LocationId locationId, LocationId neighborId, DIRECTION direction, bool biDirectional)
    {
        m_locations[locationId]->SetNeighborId(direction, neighborId);
        if (neighborId != -1)
        {
            // Set the neighbor return path
            m_locations[neighborId]->SetNeighborId(GetReverseDirection(direction), locationId);
        }
    }

    //------------------------------------------------------------------------------------------------
    std::string CGame::ReadInput()
    {
        // The default implementation is to wait for console input from stdin
        std::string inputString;
        std::getline(std::cin, inputString);
        return inputString;
    }

    //------------------------------------------------------------------------------------------------
    void CGame::WriteOutput(const char *outputString, int /*stringLength*/)
    {
        // The default implementation is to write to stdout
        std::cout << outputString;
    }

    //------------------------------------------------------------------------------------------------
    void CGame::Run()
    {
        Initialize();

        // Main loop
        for (; m_exit == false;)
        {
            BeginFrame();

            m_outputStream << "\n" << m_promptString;
            std::string commandString = ReadInput();
            m_logStream << commandString;
            m_outputStream << "\n";

            CCommand command;
            if (CCommandBuilder::Tokenize(m_dictionary, commandString, command))
            {
                CPlayer *pPlayer = GetPlayer(m_selectedPlayerId);

                // First let the player handle the command
                if(!pPlayer->DoCommand(command))
                {
                    if (!DoCommand(pPlayer, command))
                    {
                        m_outputStream << "That can't be done.\n";
                    }
                }
            }

            EndFrame();
        }

        Cleanup();
    }

    //------------------------------------------------------------------------------------------------
    bool CGame::DoCommand(CPlayer *pPlayer, const CCommand &command)
    {
        // Parse the command...
        if (command.IsMatch(WORD_HELP))
        {
            PrintHelp();
            return true;
        }

        if (command.IsMatch(WORD_EXIT))
        {
            m_exit = true;
            return true;
        }

        return false;
    }

    //------------------------------------------------------------------------------------------------
    PlayerId CGame::SelectPlayer(PlayerId playerId)
    {
        m_selectedPlayerId = playerId >= 0 && playerId < m_players.size() ? playerId : -1;
        CPlayer *pPlayer = GetPlayer(m_selectedPlayerId);

        if (pPlayer)
        {
            CLocation *pLocation = CGame::Game().GetLocation(pPlayer->GetLocationId());
            if (pLocation)
            {
                pLocation->DoLook();
            }
        }

        return m_selectedPlayerId;
    }

    //------------------------------------------------------------------------------------------------
    void CGame::PrintHelp()
    {
        m_outputStream << "No help for you yet\n";
    }

    //------------------------------------------------------------------------------------------------
    WordId CGame::MakeItem(WordId wordId, const std::string &inventoryDescription, const std::string &locationDescription)
    {
        try
        {
            m_items.emplace(wordId, std::make_unique<CItem>(wordId, inventoryDescription, locationDescription)); // throw(std::bad_alloc)
        }
        catch (std::bad_alloc &)
        {
            return 0UL - 1;
        }

        return wordId;
    }
}

File: GoNorthCS/Game.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace GoNorth
{
    //------------------------------------------------------------------------------------------------
    public class Game : CommandInterpreter
    {
        //
        // Data members
        //
        bool _exit = false;
        SortedList<int, Item> _items = new SortedList<int, Item>();
        SortedList<int, Location> _locations = new SortedList<int, Location>();
        SortedList<int, Player> _players = new SortedList<int, Player>();
        Dictionary _dictionary = new Dictionary();
        public string PromptString { get; set; }
        int _selectedPlayerId = 0;

        //
        // Constructor
        //
        public Game()
        {
            // Register words
            _dictionary.AddWord("Help", (int)WORDS.WORD_HELP);
            _dictionary.AddWord("Exit", (int)WORDS.WORD_EXIT);
            _dictionary.AddWord("Use", (int)WORDS.WORD_USE);
            _dictionary.AddWord("Go", (int)WORDS.WORD_GO);
            _dictionary.AddWord("Get", (int)WORDS.WORD_GET);
            _dictionary.AddWord("Drop", (int)WORDS.WORD_DROP);
            _dictionary.AddWord("Move", (int)WORDS.WORD_MOVE);
            _dictionary.AddWord("Look", (int)WORDS.WORD_LOOK);
            _dictionary.AddWord("North", (int)WORDS.WORD_NORTH);
            _dictionary.AddWord("Northeast", (int)WORDS.WORD_NORTHEAST);
            _dictionary.AddWord("East", (int)WORDS.WORD_EAST);
            _dictionary.AddWord("Southeast", (int)WORDS.WORD_SOUTHEAST);
            _dictionary.AddWord("South", (int)WORDS.WORD_SOUTH);
            _dictionary.AddWord("Southwest", (int)WORDS.WORD_SOUTHWEST);
            _dictionary.AddWord("West", (int)WORDS.WORD_WEST);
            _dictionary.AddWord("Northwest", (int)WORDS.WORD_NORTHWEST);
            _dictionary.AddWord("Up", (int)WORDS.WORD_UP);
            _dictionary.AddWord("Down", (int)WORDS.WORD_DOWN);
            _dictionary.AddWord("Yes", (int)WORDS.WORD_YES);
            _dictionary.AddWord("No", (int)WORDS.WORD_NO);
            _dictionary.AddWord("Inventory", (int)WORDS.WORD_INVENTORY);

            // Aliases
            _dictionary.AddWord("?", (int)WORDS.WORD_HELP);
            _dictionary.AddWord("Quit", (int)WORDS.WORD_EXIT);
            _dictionary.AddWord("Q", (int)WORDS.WORD_EXIT);
            _dictionary.AddWord("N", (int)WORDS.WORD_NORTH);
            _dictionary.AddWord("NE", (int)WORDS.WORD_NORTHEAST);
            _dictionary.AddWord("E", (int)WORDS.WORD_EAST);
            _dictionary.AddWord("SE", (int)WORDS.WORD_SOUTHEAST);
            _dictionary.AddWord("S", (int)WORDS.WORD_SOUTH);
            _dictionary.AddWord("SW", (int)WORDS.WORD_SOUTHWEST);
            _dictionary.AddWord("W", (int)WORDS.WORD_WEST);
            _dictionary.AddWord("NW", (int)WORDS.WORD_DOWN);
            _dictionary.AddWord("I", (int)WORDS.WORD_INVENTORY);
            _dictionary.AddWord("Y", (int)WORDS.WORD_YES);
            _dictionary.AddWord("N", (int)WORDS.WORD_NO);

            // Add default player and select
            _selectedPlayerId = AddPlayer(new Player());

            // Set the default prompt string
            PromptString = "> ";
        }

        //
        // Accessors
        //

        public Dictionary Dictionary
        {
            get { return _dictionary; }
        }

        //------------------------------------------------------------------------------------------------
        public Item GetItem(int WordId)
        {
            return _items[WordId];
        }

        //------------------------------------------------------------------------------------------------
        public Location GetLocation(int LocationId)
        {
            return _locations[LocationId];
        }

        //------------------------------------------------------------------------------------------------
        public Player GetPlayer(int PlayerId)
        {
            return _players[PlayerId];
        }

        //------------------------------------------------------------------------------------------------
        public static Direction WordIdToDirection(int WordId)
        {
            Direction direction = Direction.NoDirection;

            switch (WordId)
            {
                case (int) WORDS.WORD_NORTH:
                    direction = Direction.North;
                    break;
                case (int)WORDS.WORD_NORTHEAST:
                    direction = Direction.Northeast;
                    break;
                case (int)WORDS.WORD_EAST:
                    direction = Direction.East;
                    break;
                case (int)WORDS.WORD_SOUTHEAST:
                    direction = Direction.Southeast;
                    break;
                case (int)WORDS.WORD_SOUTH:
                    direction = Direction.South;
                    break;
                case (int)WORDS.WORD_SOUTHWEST:
                    direction = Direction.Southwest;
                    break;
                case (int)WORDS.WORD_WEST:
                    direction = Direction.West;
                    break;
                case (int)WORDS.WORD_NORTHWEST:
                    direction = Direction.Northwest;
                    break;
                case (int)WORDS.WORD_UP:
                    direction = Direction.Up;
                    break;
                case (int)WORDS.WORD_DOWN:
                    direction = Direction.Down;
                    break;
            }

            return direction;
        }

        //
        // Overridable Methods
        //

        //------------------------------------------------------------------------------------------------
        public virtual void Initialize() { }

        //------------------------------------------------------------------------------------------------
        public virtual void BeginFrame() { }

        //------------------------------------------------------------------------------------------------
        public virtual void EndFrame() { }

        //------------------------------------------------------------------------------------------------
        public virtual void Cleanup() { }

        //------------------------------------------------------------------------------------------------
        public virtual string ReadInput()
        {
            // The default implementation is to wait for console input from stdin
            return Console.ReadLine();
        }

        //------------------------------------------------------------------------------------------------
        public virtual void WriteOutput(string outputString)
        {
            // By default write output to the console
            Console.Write(outputString);
        }

        public virtual void WriteLog(string logString)
        {

        }

        //------------------------------------------------------------------------------------------------
        public virtual void Run()
        {
            Initialize();

            // Main loop
            for (; _exit == false;)
            {
                BeginFrame();

                WriteOutput("\n" + PromptString);
                string commandString = ReadInput();
                WriteLog(commandString);
                WriteOutput("\n");

                Command command = new Command();
                if (CommandBuilder.Tokenize(this, _dictionary, commandString, command))
                {
                    Player player = GetPlayer(_selectedPlayerId);

                    // First let the player handle the command
                    if (!player.DoCommand(this, command))
                    {
                        if (!DoCommand(this, player, command))
                        {
                            WriteOutput("That can't be done.\n");
                        }
                    }
                }

                EndFrame();
            }

            Cleanup();
        }

        //------------------------------------------------------------------------------------------------
        public override bool DoCommand(Game game, Player player, Command command)
        {
            // Parse the command...
            if (command.IsMatch((int)WORDS.WORD_HELP))
            {
                PrintHelp();
                return true;
            }

            if (command.IsMatch((int)WORDS.WORD_EXIT))
            {
                _exit = true;
                return true;
            }

            return false;
        }

        //
        // Methods
        //

        //------------------------------------------------------------------------------------------------
        public void SetLocationNeighborId(int locationId, int neighborId, Direction direction, bool biDirectional)
        {
            _locations[locationId].SetNeighborId(direction, neighborId);
            if (neighborId != -1 && biDirectional)
            {
                // Set the neighbor return path
                _locations[neighborId].SetNeighborId(GetReverseDirection(direction), locationId);
            }
        }

        //------------------------------------------------------------------------------------------------
        public int SelectPlayer(int playerId)
        {
            _selectedPlayerId = playerId >= 0 && playerId < _players.Count ? playerId : -1;
            Player player = GetPlayer(_selectedPlayerId);

            if (player != null)
            {
                Location location = GetLocation(player.LocationId);
                if (location != null)
                {
                    location.DoLook(this);
                }
            }

            return _selectedPlayerId;
        }

        //------------------------------------------------------------------------------------------------
        public void PrintHelp()
        {
            WriteOutput("No help for you yet\n");
        }

        //------------------------------------------------------------------------------------------------
        public int AddItem(Item item)
        {
            _items.Add(item.WordId, item);

            return item.WordId;
        }

        //------------------------------------------------------------------------------------------------
        public int AddLocation(Location location)
        {
            int locationId = _locations.Count;
            _locations.Add(locationId, location);
            return locationId;
        }

        //------------------------------------------------------------------------------------------------
        public int AddPlayer(Player player)
        {
            int playerId = _players.Count;
            _players.Add(playerId, player);
            return playerId;
        }

        //------------------------------------------------------------------------------------------------
        public Direction GetReverseDirection(Direction direction)
        {
            switch (direction)
            {
                case Direction.North:
                    return Direction.South;
                case Direction.Northeast:
                    return Direction.Northwest;
                case Direction.East:
                    return Direction.West;
                case Direction.Southeast:
                    return Direction.Northwest;
                case Direction.South:
                    return Direction.North;
                case Direction.Southwest:
                    return Direction.Northeast;
                case Direction.West:
                    return Direction.East;
                case Direction.Northwest:
                    return Direction.Southeast;
                case Direction.Up:
                    return Direction.Down;
                case Direction.Down:
                    return Direction.Up;
            }
            return Direction.NumDirections; // ERROR
        }
    }
}

